src/Microsoft.DocAsCode.Utility/StreamSegmentSerialization/StreamSerializer.cs

```diff
@@ -0,0 +1,209 @@
+            _writer.Write(startOffset + 4 + 4 + 1); // next
+            _writer.Write((byte)StreamSegmentType.Null);
+            return new StreamSegment(UnderlyingStream, startOffset, 4 + 4 + 1, startOffset + 4 + 4 + 1, StreamSegmentType.Null);
+        }
+
+        public StreamSegment Write(int value)
+        {
+            var startOffset = (int)UnderlyingStream.Position;
+            _writer.Write(4 + 4 + 1 + 4); // length
+            _writer.Write(startOffset + 4 + 4 + 1 + 4); // next
+            _writer.Write((byte)StreamSegmentType.Integer);
+            _writer.Write(value);
+            return new StreamSegment(UnderlyingStream, startOffset, 4 + 4 + 1 + 4, startOffset + 4 + 4 + 1 + 4, StreamSegmentType.Integer);
+        }
+
+        public StreamSegment Write(string value)
+        {
+            if (value == null)
+            {
+                return WriteNull();
+            }
+            return WriteString(value);
+        }
+
+        public StreamSegment Write(byte[] value)
+        {
+            if (value == null)
+            {
+                return WriteNull();
+            }
+            return WriteBytes(value);
+        }
+
+        public StreamSegment Write(Action<Stream> writeAction)
+        {
+            if (writeAction == null)
+            {
+                throw new ArgumentNullException(nameof(writeAction));
+            }
+            // todo : forbidden seek.
+            return WriteBytes(writeAction);
+        }
+
+        public StreamSegment Write(IReadOnlyList<object> value)
+        {
+            if (value == null)
+            {
+                return WriteNull();
+            }
+            return WriteArray(value);
+        }
+
+        public StreamSegment Write(IReadOnlyDictionary<string, object> value)
+        {
+            if (value == null)
+            {
+                return WriteNull();
+            }
+            return WriteDictionary(value);
+        }
+
+        public StreamSegment Write(object value)
+        {
+            if (value == null)
+            {
+                return WriteNull();
+            }
+            var s = value as string;
+            if (s != null)
+            {
+                return WriteString(s);
+            }
+            var bytes = value as byte[];
+            if (bytes != null)
+            {
+                return WriteBytes(bytes);
+            }
+            if (value is int)
+            {
+                return Write((int)value);
+            }
+            var list = value as IReadOnlyList<object>;
+            if (list != null)
+            {
+                return WriteArray(list);
+            }
+            var dict = value as IReadOnlyDictionary<string, object>;
+            if (dict != null)
+            {
+                return WriteDictionary(dict);
+            }
+            throw new NotSupportedException();
+        }
+
+        private StreamSegment WriteString(string value)
```

Comment:
Yes, but call different write overload.

---
In reply to: [78334947](https://github.com/dotnet/docfx/pull/687#discussion_r78334947) [](ancestors = 78334947)